### PR TITLE
feat(`anvil`): remove old ethers-dependent anvil core types

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

removes ethers usage in all of anvil. Only remaining usage is:
- tests (removal in progress)
- block subscriptions (needs further work on alloy, we will most likely roll our own version of this)

test failures will be fixed in another PR on 6808.

 Goes into #6808.

### CODE REVIEW

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 1]
```diff
@@ -2540,12 +2531,13 @@ pub fn prove_storage(
         .unwrap();
 
     let item: U256 = {
-        let decode_value = |bytes: &[u8]| rlp::decode(bytes).expect("decoding db value failed");
+        let decode_value = |mut bytes: &[u8]| {
+            <U256 as alloy_rlp::Decodable>::decode(&mut bytes).expect("decoding db value failed")
```
Comment (gakonst) [inline crates/anvil/src/eth/backend/mem/mod.rs:2535]: Doesn't `U256::decode` work?

File: crates/anvil/src/eth/error.rs [hunk 2]
```diff
@@ -278,7 +266,7 @@ pub(crate) fn decode_revert_reason(out: impl AsRef<[u8]>) -> Option<String> {
     if out.len() < SELECTOR_LEN {
         return None
     }
-    String::decode(&out[SELECTOR_LEN..]).ok()
+    DynSolType::String.abi_decode(&out[SELECTOR_LEN..]).ok().and_then(|v| v.as_str().map(|s| s.to_owned()))
```
Comment (gakonst) [inline crates/anvil/src/eth/error.rs:269]: @DaniPopes is this the best way to get the Option<String>?

File: crates/anvil/src/eth/sign.rs [hunk 3]
```diff
@@ -1,225 +0,0 @@
-use crate::eth::error::BlockchainError;
```
Comment (gakonst) [inline crates/anvil/src/eth/sign.rs:1]: @mattsse IIRC this was due to lack of object safety on signer?

**Review by gakonst — COMMENTED** [review]
getting there!

File: crates/anvil/src/eth/error.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/anvil/src/eth/error.rs:269]: No, but just use foundry_common decode_revert https://github.com/foundry-rs/foundry/blob/b51c05734d0bc8879df64d00f536acb7413c84fa/crates/evm/core/src/decode.rs#L30C8-L30C21

Comment (DaniPopes) [reply crates/anvil/src/eth/error.rs:269]: It would be String::abi_decode anyway.

File: crates/anvil/src/eth/error.rs [hunk 4]
```diff
@@ -266,10 +265,7 @@ pub(crate) fn decode_revert_reason(out: impl AsRef<[u8]>) -> Option<String> {
     if out.len() < SELECTOR_LEN {
         return None
     }
-    DynSolType::String
-        .abi_decode(&out[SELECTOR_LEN..])
-        .ok()
-        .and_then(|v| v.as_str().map(|s| s.to_owned()))
+    Some(foundry_evm::decode::decode_revert(&out[SELECTOR_LEN..], None, None))
```
Comment (DaniPopes) [inline crates/anvil/src/eth/error.rs:268]: Can just replace this whole function with `maybe_decode_revert`

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 5]
```diff
@@ -77,11 +74,11 @@ use foundry_evm::{
         },
     },
     traces::{TracingInspector, TracingInspectorConfig},
-    utils::{eval_to_instruction_result, halt_to_instruction_result, u256_to_h256_be},
+    utils::{eval_to_instruction_result, halt_to_instruction_result},
 };
 use futures::channel::mpsc::{unbounded, UnboundedSender};
 use hash_db::HashDB;
-use itertools::Itertools;
+// use itertools::Itertools;
```
Comment (onbjerg) [inline crates/anvil/src/eth/backend/mem/mod.rs:81]: nit: rm

File: crates/anvil/src/eth/backend/mem/state.rs [hunk 6]
```diff
@@ -2,41 +2,21 @@
 
 use crate::eth::{backend::db::AsHashDB, error::BlockchainError};
 use alloy_primitives::{Address, Bytes, B256, U256 as rU256};
```
Comment (onbjerg) [inline crates/anvil/src/eth/backend/mem/state.rs:4]: nit: can just refer to it using U256 now?

File: crates/anvil/tests/it/proof/mod.rs [hunk 7]
```diff
@@ -20,50 +15,52 @@ async fn can_get_proof() {
...
+        storage_root: proof.storage_hash,
+        code_hash: KECCAK_EMPTY,
     };
 
-    let rlp_account = rlp::encode(&account);
+    let rlp_account = alloy_rlp::encode(&account);
 
-    let root: H256 = api.state_root().await.unwrap().to_ethers();
+    let root: B256 = api.state_root().await.unwrap();
+    println!("proof: {:?}", proof);
```
Comment (onbjerg) [inline crates/anvil/tests/it/proof/mod.rs:36]: nit: rm

**Review by onbjerg — CHANGES_REQUESTED** [review]
pog

**Review by onbjerg — APPROVED** [review]
lgtm, tests fixed in follow up and we're merging into a feature branch so should be good to merge

File: crates/anvil/core/src/eth/serde_helpers.rs [hunk 8]
```diff
@@ -33,6 +33,35 @@ pub mod sequence {
     }
 }
 
+pub mod numeric {
```
Comment (gakonst) [inline crates/anvil/core/src/eth/serde_helpers.rs:36]: Should this be an alloy util? @DaniPopes

File: crates/anvil/src/config.rs [hunk 9]
```diff
@@ -1160,26 +1155,8 @@ impl AccountGenerator {
...
-        for idx in 0..self.amount {
-            let builder =
-                builder.clone().derivation_path(&format!("{derivation_path}{idx}")).unwrap();
-            let wallet = builder.build().unwrap().with_chain_id(self.chain_id);
-            wallets.push(wallet)
-        }
-        wallets
-    }
-
     pub fn alloy_gen(&self) -> Vec<LocalWallet> {
```
Comment (gakonst) [inline crates/anvil/src/config.rs:1158]: rename back to gen

**Pull request merged**